app/core/services/attack_service.py

```
@@ -207,6 +207,218 @@ async def duplicate_attack_service(attack_id: int, db: AsyncSession) -> AttackOu
+    """
+    from app.models.campaign import Campaign
+
+    # Verify campaign exists
+    campaign_result = await db.execute(
+        select(Campaign).where(Campaign.id == campaign_id)
+    )
+    campaign = campaign_result.scalar_one_or_none()
+    if not campaign:
+        from app.core.services.campaign_service import CampaignNotFoundError
+
+        raise CampaignNotFoundError(f"Campaign {campaign_id} not found")
```

Ruff (0.15.0)

[warning] Avoid specifying long messages outside the exception class

(TRY003)